[24기 김세준] Mission 1~4완료. Feed 컴포넌트 구현 중

kk3june opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- Mission1 사용자 입력 데이터 저장 기능 완료
- 로그인 버튼 활성화 기능 완료
- 댓글 입력 기능 완료
- 댓글 컴포넌트화, props로 데이터 전달 구현
- `추가구현사항`Feed 컴포넌트화 완료
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- scss 사용하여 스타일링 완료 (nesting)
- `Login`  e.target.value 사용하여 입력 값 전달 받아 setState로 데이터 저장 
- `Login` onClick 이벤트에 삼항연산자 사용하여 유효성 로직 구현
- `Feed` 컴포넌트화 완료, input 입력값 setState 사용하여 this.state.commentList에 배열 형태로 저장, commentList 배열 Array.map 사용하여 Comment 컴포넌트로 전달
- `Comment` 컴포넌트화 완료, 컴포넌트 내부에 댓글 HTML 구조 만든 후 `Feed` 컴포넌트에서 전달받은 데이터 props 사용하여 댓글 기능 구현

## 기타 질문 및 특이 사항
### 추가 구현 예정 사항
- `Login` 페이지와 연결될 회원가입 페이지 생성 및 백엔드 통신
- `Comment` 컴포넌트 댓글 삭제 기능 추가
-  `Feed` 컴포넌트에 commentData.json 파일 Array.map 사용하여 추가 Feed 구현(Mock data 추가 학습)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
- Main.js, Main.scss
:

Commit:
Add : Mission 1 - 사용자 입력 데이터 저장 기능 구현

Commit:
- Login 페이지 사용자 입력 데이터 저장 기능 구현
- Login 버튼 활성화 로직 구현

esphoenixc commented on src/pages/sejun/Login/Login.js:
```diff
@@ -1,10 +1,64 @@
+                pwVal: e.target.value,
+              });
+              e.preventDefault();
+            }.bind(this)}
```
onChange안에 바로 함수 작성해주신 방법은 생각 따로 못했는데 잘봤씁니다!! 저 bind(this)는 저기서 this가 undefined떠서 쓰신건가요??

younglys replied:
함수가 호출될 때의 context가 전역객체라 undefined이 출력되어 this 문맥을 잡아주기 위해 bind 함수 사용하신 것 같아요~

yeonuk-hwang replied:
arrow function 으로 사용해주는게 function + bind 메소드보다 간결하게 보입니다.

yeonuk-hwang replied:
input handler 함수들 id, pw 각기 따로 두지말고 input의 name property랑 계산된 속성명 이용해서 합칠 수 있습니다

esphoenixc commented on src/pages/sejun/Login/Login.js:
```diff
@@ -1,10 +1,64 @@
+                this.state.idVal.includes('@') && this.state.pwVal.length >= 5
+                  ? 'loginBtn active'
+                  : 'loginBtn nonActive'
+              }
```
className={}안에다가 삼항연산자 작성 깔끔하게 잘 해주신거 같습니다.

esphoenixc commented on src/pages/sejun/Login/Login.js:
```diff
@@ -1,10 +1,64 @@
+              this.setState({
+                idVal: e.target.value,
+              });
+              e.preventDefault();
```
여기서 e.preventDefault()는 어떤식으로 작동되는건가요??

YOUNGHO8762 commented on src/pages/sejun/Login/Login.js:
```diff
@@ -1,10 +1,64 @@
+                idVal: e.target.value,
+              });
+              e.preventDefault();
+            }.bind(this)}
```
function은 render() 위에 만드셔서 사용하시는 게 좋을 거 같아요!

younglys replied:
화살표 함수를 사용해 bind 함수를 생략해도 좋을 것 같습니다!

yeonuk-hwang replied:
👍 팀원들 피어리뷰 반영해주세요

younglys commented on src/pages/sejun/Login/Login.scss:
```diff
@@ -0,0 +1,49 @@
+      color: white;
+      background-color: #c4e1fb;
+      border: 1px solid #efefef;
+    }
```
CSS 속성은 연관관계를 쉽게 파악하고 유지보수를 위해 아래와 같이 권장되는 순서대로 작성해주시면 좋을 것 같습니다!
  1. Layout Properties (position, float, clear, display)
  2. Box Model Properties (width, height, margin, padding)
  3. Visual Properties (color, background, border, box-shadow)
  4. Typography Properties (font-size, font-family, text-align, text-transform)
  5. Misc Properties (cursor, overflow, z-index)

yeonuk-hwang replied:
리뷰 반영해주세요

Commit:
댓글 기능 구현중

Commit:
Merge branch 'master' of https://github.com/wecode-bootcamp-korea/24-React-Westagram-3 into feature/sejun

Commit:
Modified : 피드 컴포넌트화 완료, 댓글 기능 구현 완료

Commit:
불필요한 파일 일부 삭제

yeonuk-hwang commented:
PR 제목 양식에 맞춰주세요!

yeonuk-hwang commented on src/pages/sejun/Comment/Comment.js:
```diff
@@ -0,0 +1,20 @@
+
+class Comment extends React.Component {
+  render() {
+    const { key, user, data } = this.props;
```
변수명으로 data 처럼 범용적인 이름을 사용하는건 좋지 않습니다. 의미가 드러나는 변수명으로 사용해주세요

yeonuk-hwang commented on src/pages/sejun/Comment/Comment.js:
```diff
@@ -0,0 +1,20 @@
+    const { key, user, data } = this.props;
+    return (
+      <div className="comment">
+        <span className="commentId" id={key}>
```
- `id` 보다는 `className` 을 활용해주시기 바랍니다.
- HTML에서  `id` 는 문서 내에서 유일해야 합니다. 이러한 이유로 단일 요소를 다루는 경우 `id`를 사용할 수 있습니다.
- 하지만 정말로 해당 요소에 대한 이벤트나 스타일 핸들링이 프로젝트 전체에서 단 한번만 사용 되는지 충분히 고민해야합니다. 일반적으로 컴포넌트는 재사용 될 수 있기 때문에 같은 요소가 여러번 생성될 수 있습니다. 이러한 이유로 보통의 경우에는 `className` 을 사용합니다.
- 성능, 혹은 기타 이유로 `id` 를 반드시 사용해야 하는 경우가 생긴다면 충분히 unique한 네이밍을 해주셔야 합니다.

yeonuk-hwang commented on src/pages/sejun/Comment/Comment.js:
```diff
@@ -0,0 +1,20 @@
+        </span>
+        <span className="commentContent">{data}</span>
+        <button className="likeBtn">
+          <i className="far fa-heart"></i>
```
self-closing 적용해주세요

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.js:
```diff
@@ -0,0 +1,132 @@
+    this.setState({
+      comment: '',
+      commentList: this.state.commentList.concat(this.state.comment),
+      key: this.state.uuid + 1,
```
state안에 uuid는 없는 것 같습니다.

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.js:
```diff
@@ -0,0 +1,132 @@
+
+class Feed extends React.Component {
+  state = {
+    key: 1,
```
key는 UI에 사용되는 값이 아닌 것 같네요

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.js:
```diff
@@ -0,0 +1,132 @@
+class Feed extends React.Component {
+  state = {
+    key: 1,
+    id: '3_jun',
```
[공식문서 React로 사고하기](https://ko.reactjs.org/docs/thinking-in-react.html#step-3-identify-the-minimal-but-complete-representation-of-ui-state) 를 보시면 어떤 값들이 state가 되어야하는지에 대해 적혀있습니다.

- id는 변하지 않는 값 같네요

> 각각 살펴보고 어떤 게 state가 되어야 하는 지 살펴봅시다. 이는 각 데이터에 대해 아래의 세 가지 질문을 통해 결정할 수 있습니다

> 1. 부모로부터 props를 통해 전달됩니까? 그러면 확실히 state가 아닙니다.

> 2. 시간이 지나도 변하지 않나요? 그러면 확실히 state가 아닙니다.

> 3. 컴포넌트 안의 다른 state나 props를 가지고 계산 가능한가요? 그렇다면 state가 아닙니다.

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.js:
```diff
@@ -0,0 +1,132 @@
+            <div className="articleAuthor">
+              <img
+                src="https://yt3.ggpht.com/ytc/AKedOLSJf_PYHF9czwJ0c99ARvsOkYLzoUGXSVhvOvlAkoc=s900-c-k-c0x00ffffff-no-rj"
+                alt=""
```
alt 속성 반드시 채워주세요

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.js:
```diff
@@ -0,0 +1,132 @@
+                    <button className="likeBtn">
+                      <i className="far fa-heart"></i>
+                    </button>
+                  </div>
```
반복되는 UI는 map을 이용해서 표현해보세요

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.js:
```diff
@@ -0,0 +1,132 @@
+                  </div>
+                  {this.state.commentList.map(item => (
+                    <Comment
+                      key={this.state.key}
```
이러면 모든 코멘트가 같은 key 값을 가질 것 같네요 state.key는 하니까요

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.scss:
```diff
@@ -0,0 +1,140 @@
+.contentWrapper {
```
다른 요소에 영향을 미치는 것을 방지하기 위해 최상위에 네스팅하는 요소는 일반적으로 컴포넌트 이름과 동일하게 지어주는게 좋습니다

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.scss:
```diff
@@ -0,0 +1,140 @@
+.contentWrapper {
+  display: flex;
+  justify-content: space-between;
```
자주 사용되는 속성들은 mixin등 활용해보세요

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.scss:
```diff
@@ -0,0 +1,140 @@
+  justify-content: space-between;
+  width: 975px;
+  margin: auto;
+  main {
```
다른 선택자들끼리 한 줄 띄워주세요

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.scss:
```diff
@@ -0,0 +1,140 @@
+      }
+    }
+
+    article {
```
- tag selector는 다른 요소들에 영향을 받을, 미칠 여지가 너무 많기 때문에 추후 유지보수에 악영향을 미칠 수 있습니다.
- 스타일링을 줄 때는 가급적 요소마다 className을 부여하셔서 스타일링을 주시는게 좋습니다.

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.scss:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    article {
+      width: 100%;
```
이미 display:block 요소인데 width:100% 줄 필요 없습니다

yeonuk-hwang commented on src/pages/sejun/Feed/Feed.scss:
```diff
@@ -0,0 +1,140 @@
+          }
+          .comments {
+            padding: 10px 10px;
+            .commentWrapper {
```
스타일링이 필요없다면 굳이 클래스네임을 부여할 이유도, 네스팅을 감쌀 이유도 없습니다!

yeonuk-hwang commented on src/pages/sejun/Login/Login.js:
```diff
@@ -1,10 +1,80 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      idVal: '',
```
idValue로 수정해주세요 변수명은 명료할수록 좋습니다

yeonuk-hwang commented on src/pages/sejun/Login/Login.js:
```diff
@@ -1,10 +1,80 @@
+        nickname: 'Hope',
+        password: this.state.pwVal,
+      }),
+    }).then(res => res.json());
```
res.json을 한 결과를 사용안하면 res.json 할 필요 없어보입니다

yeonuk-hwang commented on src/pages/sejun/Login/Login.js:
```diff
@@ -1,10 +1,80 @@
     return (
-      <div>
-        <h3>로그인</h3>
+      <div className="loginBoard">
```
최상위 태그의 className은 컴포넌트 이름과 동일하게 해주세요, 다만 현재 팀프로젝트에서는 겹칠 염려가 있으니 Login{이름} 이런식으로 해도 무방해보이네요

yeonuk-hwang replied:
ex) LoginYeonuk

yeonuk-hwang commented on src/pages/sejun/Login/Login.scss:
```diff
@@ -0,0 +1,49 @@
+  }
+}
+
+.forgetPw {
+  display: block;
+  margin-top: 150px;
+  text-align: center;
+  color: #003569;
+  font-weight: 600;
+}
```
네스팅에 포함시켜주세요

yeonuk-hwang commented on src/pages/sejun/Main/Main.js:
```diff
@@ -1,10 +1,173 @@
 import React from 'react';
+// import Comment from '../Comment/Comment';
```
불필요한 주석은 삭제해주세요

yeonuk-hwang commented on src/pages/sejun/Main/Main.js:
```diff
@@ -1,10 +1,173 @@
+                alt=""
+              />
+              <div className="stryId">wkndshi</div>
+            </div>
```
반복되는 UI는 Array.map을 이용해서 표현해보세요 이렇게 정적인 컨텐츠는 상수데이터와 조합하면 되어보입니다

yeonuk-hwang commented on src/pages/sejun/Main/Main.scss:
```diff
@@ -0,0 +1,218 @@
+body {
+  background-color: #fafafa;
+}
```
모든 페이지의 스타일링에 영향을 미치는 속성입니다 지양해주시고 필요한 경우에는 컴포넌트의 scss 파일이 아닌 common.scss에 있는게 옳아보입니다.

yeonuk-hwang reviewed (commented):
리뷰 확인 및 반영해주세요 세준님

Commit:
- 범용적인 data명 수정( Comment 컴포넌트 data -> comment)
- Comment 컴포넌트 id 속성 삭제 ( 추후 key props 에러 수정 예정)
- i tag self-closing 완료
- this.state 변하지 않는 속성 및 사용하지 않는 속성 제거 (Feed 컴포넌트 id,uuid)
- 비어있는 alt 속성 수정 완료
- Feed 컴포넌트에서 Comment 컴포넌트로 전달하는 key 값 제거 완료 (추후 각 comment id 값 추가 예정)
- mixin 기능 사용
- scss 지적사항 수정 완료
- Feed 컴포넌트 Article 컴포넌트로 변경(추후 Feed 컴포넌트 생성 예정)
